File: TrolleyCore/Networking/JSON/TRLJSON.m
//
//  TRLJSON.m
//  TrolleyNetworkingTools
//
//  Created by Harry Wright on 05.09.17.
//  Copyright © 2017 Off-Piste.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

#import "TRLJSON.h"
//#import "TRLJSON+Bridge.h"

#import "TRLMutableJSON.h"
#import "TRLJSONBase_Dynamic.h"

#import "TRLMutableDictionary.h"
#import "TRLMutableArray.h"

#import "NSArray+Map.h"
#import <TrolleyCore/TrolleyCore-Swift.h>

@implementation TRLJSON

+ (instancetype)bridge:(TRLJSONBase *)superclass {
    return [self object:superclass.rawValue];
}

- (id)object {
    return self.rawValue;
}

- (NSUInteger)count {
    switch (self.rawType) {
        case JSONTypeArray: return self->rawArray.count;
        case JSONTypeDictionary: return self->rawDictionary.count;
        default: return 0;
    }
}

+ (instancetype)object:(id)obj {
    if ([obj isKindOfClass:[NSArray<TRLJSON *> class]]) {
        NSArray *array = (NSArray<TRLJSON *>*)obj;
        if (array.isJSON) {
            return [[TRLJSON alloc] initWithArray:(NSArray<TRLJSON *> *)obj];
        }
    } else if ([obj isKindOfClass:[NSDictionary class]]) {
        NSDictionary *dict = (NSDictionary *)obj;
        if (dict.isJSON) {
            return [[TRLJSON alloc] initWithDictionary:(NSDictionary<NSString *, TRLJSON *>*)obj];
        }
    }
    return [[TRLJSON alloc] initWithRawValue:obj];
}

+ (instancetype)null {
    return [TRLJSON object:ObjcNull];
}

+ (instancetype)parseJSON:(NSString *)jsonString {
    NSData *d = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    if (d) {
        return [TRLJSON object:d];
    }
    return [TRLJSON null];
}

- (instancetype)initWithArray:(NSArray<TRLJSON *> *)array {
    return [TRLJSON object:[array map:^id(TRLJSON * obj, NSUInteger idx) {
        return [TRLJSON object:obj.object];
    }]];
}

- (instancetype)initWithDictionary:(NSDictionary<NSString *, TRLJSON *>*)dictionary {
    NSMutableDictionary *newDictionary = @{}.mutableCopy;
    [dictionary enumerateKeysAndObjectsUsingBlock:^(NSString *key, TRLJSON *obj, BOOL *stop) {
        newDictionary[key] = obj.object;
    }];
    return [TRLJSON object:newDictionary];
}

- (id)copyWithZone:(NSZone *)zone {
    TRLJSON *copy = [[[self class] alloc] initWithRawValue:ObjcNull];
    TRLJSONBaseSetRawValue(copy, [self.rawValue copyWithZone:zone]);
    return copy;
}

- (id)mutableCopyWithZone:(NSZone *)zone {
    TRLMutableJSON *copy = [[TRLMutableJSON allocWithZone:zone] initWithRawValue:ObjcNull];
    copy.object = [self.rawValue copyWithZone:zone];
    return copy;
}

- (instancetype)objectForKeyedSubscript:(NSString *)key {
    return [TRLJSON bridge:[super objectForKeyedSubscript:key]];
}

- (instancetype)objectAtIndexedSubscript:(NSUInteger)idx {
    return [TRLJSON bridge:[super objectAtIndexedSubscript:idx]];
}

- (NSArray<TRLJSON *> *)array {
    if (self.rawType == JSONTypeArray) {
        return [self->rawArray map:^id(id obj, NSUInteger idx) {
            return [TRLJSON object:obj];
        }].array;
    }
    return nil;
}

- (NSArray<TRLJSON *> *)arrayValue {
    return self.array ? self.array : @[];
}

- (NSDictionary<NSString *,TRLJSON *> *)dictionary {
    if (self.rawType == JSONTypeDictionary) {
        NSMutableDictionary<NSString *,TRLJSON *> *d = @{}.mutableCopy;
        [self->rawDictionary enumerateKeysAndObjectsUsingBlock:^(NSString *key, id obj, BOOL *stop) {
            [d setValue:[TRLJSON object:obj] forKey:key];
        }];
        return d;
    }
    return nil;
}

- (NSDictionary<NSString *, TRLJSON *> *)dictionaryValue {
    return self.dictionary ? self.dictionary : @{};
}

- (NSNumber *)number {
    switch (self.rawType) {
        case JSONTypeNumber: return self->rawNumber;
        case JSONTypeBool: return @(self->rawBool);
        case JSONTypeString: {
            NSNumber *n = [[NSNumber alloc] initWithString:self->rawString];
            if (n) {
                return n;
            }
            return nil;
        }
        default: return nil;
    }
}

- (NSNumber *)numberValue {
    return self.number ? self.number : 0;
}

- (NSString *)string {
    switch (self.rawType) {
        case JSONTypeString: return self->rawString;
        case JSONTypeNumber: return self->rawNumber.stringValue;
        case JSONTypeBool: return self->rawBool ? @"true" : @"false";
        default: return nil;
    }
}

- (NSString *)stringValue {
    return self.string ? self.string : @"";
}

@end

File: TrolleyCore/TRLJSON.swift
/////////////////////////////////////////////////////////////////////////////////
//
//  TRLJSON.swift
//  TrolleyCore
//
//  Created by Harry Wright on 24.08.17.
//  Copyright © 2017 Off-Piste.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

import Foundation

/// TRLJSON is a SwiftyJSON/Objc port
/// everything runs though SwiftJSON
/// but is designed to be an interface
/// for my Objc Tools
public class TRLJSON: NSObject {

    public internal(set) var json: JSON

    // MARK: Inits

    public override init() {
        self.json = JSON.null
    }

    public init(data: Data) {
        self.json = JSON(data: data)
    }

    public init(nullable_data: Data?) throws {
        if let data = nullable_data {
            self.json = JSON(data: data)
        } else {
            throw NSError(domain: ErrorDomain, code: ErrorInvalidJSON, userInfo: [NSLocalizedDescriptionKey:"Invalid JSON"])
        }
    }

    public init(_ objects: Any) {
        self.json = JSON(objects)
    }

    internal init(json: JSON) {
        self.json = json
    }

    final public var type: JSONType {
        return self.json.type
    }

    // MARK: Objc Overrides

    public override var description: String {
        return json.description
    }

    public override func isEqual(_ object: Any?) -> Bool {
        if let json = object as? JSON {
            return json == self.json
        } else if let json = object as? TRLJSON {
            return json.json == self.json
        }
        return super.isEqual(object)
    }

    final public override func value(forKey key: String) -> Any? {
        return self.json[key].object
    }

    final public override func mutableCopy() -> Any {
        return TRLMutableJSON(json: self.json)
    }
}

// MARK: - Get Properties
extension TRLJSON {

    /// <#Description#>
    ///
    /// - Parameter key: <#key description#>
    final public subscript(key: String) -> TRLJSON {
        return TRLJSON(json: self.json[key])
    }

    /// An error with the JSON. Will be nil
    /// if the JSON is valid
    final public var error: Error? {
        return self.json.error
    }

    /// The `Dictionary` value for the current JSON key
    ///
    /// - note: Will be empty if the JSON
    ///         value is not a dictionary
    final public var dictionary: [String: Any] {
        return self.json.dictionaryObject ?? [:]
    }

    /// The `String` value for the current JSON key
    ///
    /// - note: Will be set to `""` if the JSON
    ///         value is not a string, or string
    ///         convertable object
    final public var string: String {
        return self.json.stringValue
    }

    /// The `Bool` value for the current JSON key
    ///
    /// - note: Will be set to false if the JSON
    ///         value is not a BOOL
    @objc(boolean) final public var bool: Bool {
        return self.json.boolValue
    }

    /// The `NSNumber` value for the current JSON key
    ///
    /// - note: Will be set to `0` if the JSON
    ///         value is not a NSNumber or number
    ///         type.
    final public var number: NSNumber {
        return self.json.numberValue
    }

    /// The `Array<Any>` value for the current JSON key
    ///
    /// - note: Will be set to empty if the JSON
    ///         value is not an Arrat
    final public var array: [Any] {
        return self.json.arrayObject ?? []
    }

    /// The `Raw String` of the current JSON, this is
    /// helpful when wanting to view the JSON structure
    ///
    /// - note: Will be set to `""` is the JSON
    ///         is not valid
    final public var rawString: String {
        return self.json.rawString() ?? ""
    }

    /// The `Raw Data` for the current JSON
    ///
    /// - note: Will be `nil` if the `Data` is not valid
    final public var rawData: Data? {
        return try? self.json.rawData()
    }

}

// MARK: - NSObject Depreciations
extension TRLJSON {

    @available(*, unavailable, message: "TRLJSON is immutable")
    public override func setValue(_ value: Any?, forKey key: String) { }

    @available(*, unavailable, message: "TRLJSON is immutable")
    public override func setValue(_ value: Any?, forKeyPath keyPath: String) { }

    @available(*, unavailable, message: "TRLJSON is immutable")
    public override func setValue(_ value: Any?, forUndefinedKey key: String) { }

    @available(*, unavailable, message: "TRLJSON is immutable")
    public override func setNilValueForKey(_ key: String) { }

    @available(*, unavailable, message: "TRLJSON is immutable")
    public override func setValuesForKeys(_ keyedValues: [String : Any]) { }

}

// MARK: - Converting JSON -> TRLJSON
extension ObjectiveCSupport {

    public static func convert(object: TRLJSON) -> JSON {
        return object.json
    }

    public static func convert(object: JSON) -> TRLJSON {
        return TRLJSON(json: object)
    }

}

// MARK: - Request+JSON
extension Request {

    @nonobjc public func responseJSON(callback: @escaping (JSON?, Error?) -> Void) {
        self.response { (data, error) in
            if let error = error {
                callback(nil, error)
            } else {
                let json = JSON(data: data!)
                if let json_error = json.error {
                    callback(nil, json_error)
                } else {
                    callback(json, nil)
                }
            }
        }
    }

}

public typealias ArrayBlock = (Any, Int, Bool) -> Void

public typealias DictionaryBlock = (String, Any, Bool) -> Void

// MARK: - Objective C enumerations
extension TRLJSON {

    /// Applies a given block object to the entries of the dictionary.
    ///
    /// If the block sets *stop to true, the enumeration stops.
    ///
    /// - warning: If the JSON is not a dictionary then the
    ///            block will not be called.
    ///
    /// - Parameter block: A block object to operate on entries in the dictionary.
    final public func enumerateKeysAndObjectsUsingBlock(_ block: DictionaryBlock) {
        if self.json.type == .dictionary {
            (self.dictionary as NSDictionary).enumerateKeysAndObjects({ (key, obj, stop) in
                if let key = key as? String {
                    block(key, obj, stop.pointee.boolValue)
                }
            })
        }
        return
    }

    /// Executes a given block using each object in the array, 
    /// starting with the first object and continuing through the 
    /// array to the last object.
    ///
    /// - warning: If the JSON is not an array then the
    ///            block will not be called.
    ///
    /// - Parameter block: The block to apply to elements in the array.
    final public func enumerateObjects(_ block: ArrayBlock) {
        if self.json.type == .array {
            (self.array as NSArray).enumerateObjects({ (obj, index, stop) in
                block(obj, index, stop.pointee.boolValue)
            })
        }
        return
    }

}
